tket/src/WeightSubgrMono/Searching/ValueOrdering.cpp

```diff
@@ -0,0 +1,117 @@
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+/*
+We are using "solution biased searching" as our heuristic. See the paper
+"Sequential and Parallel Solution-Biased Search for Subgraph Algorithms".
+The idea is, rather than always mapping into the target vertex with largest
+possible degree, we should try lower degrees occasionally.
+This is done simply by choosing a random vertex,
+but letting the probability of choosing a vertex depend on the degree,
+so that lower degrees are less likely.
+*/
+
+ValueOrdering::ValueOrdering() {
+  // 2^{-5} = 1/32 ~ 0.03. A choice with probability <3% is probably
+  // not very significant for the overall search.
+  m_data.resize(5);
+  m_data.back().mass = 1;
+  for (unsigned index = m_data.size() - 1; index > 0; --index) {
+    m_data[index - 1].mass = 2 * m_data[index].mass;
+  }
+}
+
+void ValueOrdering::fill_data(
+    const std::set<VertexWSM>& possible_values,
+    const NeighboursData& target_ndata) {
+  // Multipass algorithm; simple and efficient enough.
+  // Pass one: find the max degree.
+  size_t max_degree = 0;
+  for (VertexWSM tv : possible_values) {
+    max_degree = std::max(max_degree, target_ndata.get_degree(tv));
+  }
+  TKET_ASSERT(max_degree > 0);
+
+  // Pass two: record all those vertices with large enough degree.
+  for (auto& entry : m_data) {
+    entry.vertices.clear();
+  }
+  for (VertexWSM tv : possible_values) {
+    const auto degree = target_ndata.get_degree(tv);
+    if (degree + m_data.size() > max_degree) {
+      const auto index = max_degree - degree;
+      m_data[index].vertices.push_back(tv);
+    }
+  }
+  TKET_ASSERT(!m_data[0].vertices.empty());
+}
+
+VertexWSM ValueOrdering::get_random_choice_from_data(RNG& rng) const {
+  // We need probability proportional to the mass; so get the total mass.
+  // Consider each vertex as having k copies, k corresponding to the mass,
+  // list them all, and choose one uniformly.
+
+  std::size_t mass_sum = 0;
+  for (const auto& entry : m_data) {
+    mass_sum += entry.vertices.size() * entry.mass;
+  }
+  TKET_ASSERT(mass_sum > 0);
+  const auto index = rng.get_size_t(mass_sum - 1);
+
+  // Now, choose the vertex corresponding to this index.
+  std::size_t preceding_mass = 0;
+  for (const auto& entry : m_data) {
+    std::size_t mass_after_these_vertices =
+        preceding_mass + entry.vertices.size() * entry.mass;
+    if (mass_after_these_vertices < index) {
+      preceding_mass = mass_after_these_vertices;
+      continue;
+    }
+    for (auto vertex : entry.vertices) {
+      size_t mass_after_this_vertex = preceding_mass + entry.mass;
+      if (mass_after_this_vertex >= index) {
+        return vertex;
+      }
+      preceding_mass = mass_after_this_vertex;
+    }
+  }
+
+  // It's an error if we reach here, although a "harmless" one:
+  // it just means our calculation of the solution biased heuristic is wrong.
+  // This assert can just be removed in an "emergency" until the bug is sorted.
```

Comment:
If it ever DID hit that code, it would be a bug. But it hasn't in thousands of tested problems, so it's very unlikely that there is a bug. (Certainly it has been exercised tens of thousands of times, and with usually < 100 vertices you might expect the chance of an edge case leading to such a bug to be >0.0001, so you WOULD expect it to happen at least once.

But, even if it IS a bug, it is annoying but harmless; as I say, in an "emergency" (someone complaining that they can't run and get a solution, and they need one NOW...) you could just comment it out and use it, without waiting for a fix.